contracts/libraries/PositionValue.sol

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity >=0.7.0;
+
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+import '@uniswap/v3-core/contracts/libraries/FixedPoint128.sol';
+import '@uniswap/v3-core/contracts/libraries/TickMath.sol';
+import '../interfaces/INonfungiblePositionManager.sol';
+import './LiquidityAmounts.sol';
+import './PoolAddress.sol';
+import './PositionKey.sol';
+
+library PositionValue {
+    function principal(INonfungiblePositionManager nft, uint256 tokenId)
+        internal
+        view
+        returns (uint256 amount0, uint256 amount1)
+    {
+        (, , address token0, address token1, uint24 fee, int24 tickLower, int24 tickUpper, uint128 liquidity, , , , ) =
+            nft.positions(tokenId);
+
+        (uint160 sqrtRatioX96, , , , , , ) =
+            IUniswapV3Pool(
+                PoolAddress.computeAddress(
+                    nft.factory(),
+                    PoolAddress.PoolKey({token0: token0, token1: token1, fee: fee})
+                )
+            )
+                .slot0();
+
+        return
+            LiquidityAmounts.getAmountsForLiquidity(
+                sqrtRatioX96,
+                TickMath.getSqrtRatioAtTick(tickLower),
+                TickMath.getSqrtRatioAtTick(tickUpper),
+                liquidity
+            );
+    }
+
+    function fees(INonfungiblePositionManager nft, uint256 tokenId)
+        internal
+        view
+        returns (uint256 amount0, uint256 amount1)
+    {
+        (
+            ,
+            ,
+            address token0,
+            address token1,
+            uint24 fee,
+            int24 tickLower,
+            int24 tickUpper,
+            uint128 liquidity,
+            uint256 positionFeeGrowthInside0LastX128,
+            uint256 positionFeeGrowthInside1LastX128,
+            ,
+
+        ) = nft.positions(tokenId);
+
+        (uint256 poolFeeGrowthInside0LastX128, uint256 poolFeeGrowthInside1LastX128) =
+            getFeeGrowthInside(nft, token0, token1, fee, tickLower, tickUpper);
+
+        amount0 = FullMath.mulDiv(
+            poolFeeGrowthInside0LastX128 - positionFeeGrowthInside0LastX128,
+            liquidity,
+            FixedPoint128.Q128
+        );
+
+        amount1 = FullMath.mulDiv(
+            poolFeeGrowthInside1LastX128 - positionFeeGrowthInside1LastX128,
+            liquidity,
+            FixedPoint128.Q128
+        );
+    }
+
+    function getFeeGrowthInside(
```

Comment:
you want the current fee growth inside for the tick range, not the last time the nft's position was poked (e.g. via an nft burn or mint at the same tick range)

i'm pretty sure this requires computing this

https://github.com/Uniswap/uniswap-v3-core/blob/b2c5555d696428c40c4b236069b3528b2317f3c1/contracts/libraries/Tick.sol#L60-L95

Reply:
you'll want to do it more like this so you can avoid reading the fee growth globals if you're above/below the tick range
https://github.com/Uniswap/uniswap-v3-core/blob/b2c5555d696428c40c4b236069b3528b2317f3c1/contracts/UniswapV3Pool.sol#L202-L233